File: leetcode/easy/941.Valid Mountain Array/js/validMountainArray.js
"use strict";
/**
 * @param {number[]} arr
 * @return {boolean} if the arr is a mountain array
 */

// 1. edge cases arr length >= 3
// 2. var isItDecreasing = false, isItAscending = false
// 3. for from i + 1 until arr length;
// 3.1. if arr[i] > arr[i - 1]... isItAscending = true
// 3.2. else if arr[i] < arr[i - 1]... isItDescending = true
// 3.3. if isItDescending === true && arr[i] >= arr[i - 1]... return false
// 4. if isItAscending === false... return false
// 5. return true

// My Approach
var validMountainArray = function(arr) {
    if (arr.length < 3) {
        return false;
    } else {
        let isItDecreasing = false;
        let isItAscending = false;
        for (let i = 1; i < arr.length; i++) {
            if (arr[i] > arr[i - 1]) {
                isItAscending = true;
            } else if (arr[i] < arr[i -1]) {
                isItDecreasing = true;
            } else if(arr[i] === arr[i - 1]) {
                return false;
            }
            if (arr[i] > arr[i - 1] && isItDecreasing === true) {
                return false;
            };
        };
        if (isItDecreasing === false || isItAscending === false) {
            return false;
        };
        return true;
    };
};

// Testing
let nums1 = [2,1];
let nums2 = [3,5,5];
let nums3 = [0,3,2,1];
console.log(validMountainArray(nums3))

// Approach 1: One Pass
var validMountainArray = function(arr) {
    let i = 0;
    while (i + 1 < arr.length && arr[i] < arr[i + 1]) {
        i++;
        console.log(i);
    };
    if (i === 0 || i === arr.length -1) {
        return false;
    };
    console.log(arr[i], " ", arr[i + 1]);
    while (i + 1 < arr.length && arr[i] > arr[i + 1]) {
        i++
        console.log(i);
    };
    return i === arr.length -1;
};
// Testing
let dataTest0 = [0,3,2,1];
console.log(validMountainArray(dataTest0));
/**
 * Complexity Analysis

Time Complexity: O(N), where N is the length of A.

Space Complexity: O(1).
 */

File: leetcode/easy/941.Valid Mountain Array/python/validMountainArray.py
"""Valid Mountain Array"""
# My Approach
def valid_mountain_array(arr):
    if len(arr) < 3:
        return False
    greater_number = max(arr)
    index_greater_number = arr.index(greater_number)
    increased_validations = index_greater_number
    decreased_validation = len(arr) - 1 - index_greater_number
    if decreased_validation == 0 or increased_validations == 0:
        return False
    count_increased = count_decreased = 0
    for j in range(len(arr) - 1):
        if arr[j] < arr[j + 1] and j < index_greater_number:
            count_increased += 1
        elif arr[j] > arr[j + 1] and j >= index_greater_number:
            count_decreased += 1
        else:
            return False
    if count_increased == increased_validations and count_decreased == decreased_validation:
        return True

test1 =  [0,1,2,3,4,5,6,7,8,9]
print(valid_mountain_array(test1))
l = test1[:]
print(l)
m = max(l)
print(m)
print(l.index(m))
#1. input: integers array
#   output: boolean according if the integer nums strictly increase and then decrease

#2. edge cases: if the length it's >= 3 
#   iterate the array
#   conditional statement if numbers are increasing and after decreasing

# 3. if lenght arr >= 3: then return False
#     for loop know what is the index of the element greater
#     for loop each element range
#     if arr[i] < arr[i + 1] and i < greater: continue
#     elif arr[i] > arr[i + 1] and i >= than greater: continue
#     else return False
#     if all the tests pass so now return True

# 4. coding you can find the greater with max function 

# Approach 1: One Pass
def valid_mountain_array(arr):
    i = 0
    while i + 1 < len(arr) and arr[i] < arr[i + 1]:
        i += 1
    if i == 0 or i == len(arr) -1:
        return False
    while i + 1 < len(arr) and arr[i] > arr[i + 1]:
        i += 1
    return i == len(arr) - 1

# Testing 
dataTest0 = [2,1]
dataTest1 = [3,5,5]
dataTest2 = [0,3,2,1]
dataTest3 = [0,2,3,4,5,2,1,0]
dataTest4 = [0,2,3,4,5,2]
dataTest5 = [9,8,7,6,5,4,3,2,1,0]
dataTest6 = [2,1,2,3,5,7,9,10,12,14,15,16,18,14,13]

print(valid_mountain_array(dataTest6))

'''Complexity Analysis

Time Complexity: O(N), where N is the length of A.

Space Complexity: O(1).'''